modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuilder.java

```diff
@@ -0,0 +1,273 @@
+import static java.util.stream.Collectors.toList;
+import static org.apache.ignite.internal.util.IgniteUtils.shutdownAndAwaitTermination;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.raft.Peer;
+import org.apache.ignite.internal.raft.service.RaftGroupService;
+import org.apache.ignite.internal.schema.configuration.index.TableIndexView;
+import org.apache.ignite.internal.storage.MvPartitionStorage;
+import org.apache.ignite.internal.storage.RowId;
+import org.apache.ignite.internal.table.TableImpl;
+import org.apache.ignite.internal.table.distributed.TableMessagesFactory;
+import org.apache.ignite.internal.table.distributed.command.BuildIndexCommand;
+import org.apache.ignite.internal.thread.NamedThreadFactory;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+import org.apache.ignite.network.ClusterService;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Class for managing the index building process.
+ */
+class IndexBuilder {
+    private static final IgniteLogger LOG = Loggers.forClass(IndexBuilder.class);
+
+    /** Batch size of row IDs to build the index. */
+    private static final int BUILD_INDEX_ROW_ID_BATCH_SIZE = 100;
+
+    /** Message factory to create messages - RAFT commands. */
+    private static final TableMessagesFactory TABLE_MESSAGES_FACTORY = new TableMessagesFactory();
+
+    /** Busy lock to stop synchronously. */
+    private final IgniteSpinBusyLock busyLock;
+
+    /** Cluster service. */
+    private final ClusterService clusterService;
+
+    /** Index building executor. */
+    private final ExecutorService buildIndexExecutor;
+
+    IndexBuilder(String nodeName, IgniteSpinBusyLock busyLock, ClusterService clusterService) {
+        this.busyLock = busyLock;
+        this.clusterService = clusterService;
+
+        int cpus = Runtime.getRuntime().availableProcessors();
+
+        buildIndexExecutor = new ThreadPoolExecutor(
+                cpus,
+                cpus,
+                30,
+                TimeUnit.SECONDS,
+                new LinkedBlockingQueue<>(),
+                NamedThreadFactory.create(nodeName, "build-index", LOG)
+        );
+    }
+
+    /**
+     * Stops the index builder.
+     */
+    void stop() {
+        shutdownAndAwaitTermination(buildIndexExecutor, 10, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Initializes the build of the index.
+     */
+    void startIndexBuild(TableIndexView tableIndexView, TableImpl table) {
+        for (int partitionId = 0; partitionId < table.internalTable().partitions(); partitionId++) {
+            buildIndexExecutor.submit(new BuildIndexTask(table, tableIndexView, partitionId, null));
```

Comment:
Please pass the real row id value instead of null.

Reply:
At this point, we do not know for sure whether the partition will be on this node or not, so we need to make sure that we will build the index and only then use the index storage.

Reply:
How does that make sense? By the time tasks starts, partition should already be available, right? How can it not be available at the time of task creation?

Reply:
Discussed personally, I will create TODO.